packages/devtools_app/integration_test/test_infra/test_app_driver.dart

```diff
@@ -0,0 +1,244 @@
+  /// of app.
+  final String testAppPath;
+
+  late Process? runProcess;
+
+  late int runProcessId;
+
+  final _allMessages = StreamController<String>.broadcast();
+
+  Uri get vmServiceUri => _vmServiceWsUri;
+  late Uri _vmServiceWsUri;
+
+  Future<void> startProcess();
+
+  Future<void> start() async {
+    await startProcess();
+    assert(
+      runProcess != null,
+      '\'runProcess\' cannot be null. Assign \'runProcess\' inside the '
+      '\'startProcess\' method.',
+    );
+
+    // This class doesn't use the result of the future. It's made available
+    // via a getter for external uses.
+    unawaited(
+      runProcess!.exitCode.then((int code) {
+        _debugPrint('Process exited ($code)');
+      }),
+    );
+
+    listenToProcessOutput(runProcess!, printCallback: _debugPrint);
+
+    // Stash the PID so that we can terminate the VM more reliably than using
+    // proc.kill() (because proc is a shell, because `flutter` is a shell
+    // script).
+    final Map<String, dynamic> connected =
+        await waitFor(event: 'daemon.connected');
+    runProcessId = connected['params']['pid'];
+
+    // Set this up now, but we don't wait it yet. We want to make sure we don't
+    // miss it while waiting for debugPort below.
+    final Future<Map<String, dynamic>> started =
+        waitFor(event: 'app.started', timeout: _appStartTimeout);
+
+    final Map<String, dynamic> debugPort =
+        await waitFor(event: 'app.debugPort', timeout: _appStartTimeout);
+    final String wsUriString = debugPort['params']['wsUri'];
+    _vmServiceWsUri = Uri.parse(wsUriString);
+
+    // Map to WS URI.
+    _vmServiceWsUri =
+        convertToWebSocketUrl(serviceProtocolUrl: _vmServiceWsUri);
+
+    await started;
+  }
+
+  Future<int> killGracefully() async {
+    _debugPrint('Sending SIGTERM to $runProcessId..');
+    Process.killPid(runProcessId);
+
+    final killFuture =
+        runProcess!.exitCode.timeout(_quitTimeout, onTimeout: _killForcefully);
+    unawaited(_killAndShutdown(killFuture));
+    return killFuture;
+  }
+
+  Future<int> _killForcefully() async {
+    _debugPrint('Sending SIGKILL to $runProcessId..');
+    Process.killPid(runProcessId, ProcessSignal.sigkill);
+
+    final killFuture = runProcess!.exitCode;
+    unawaited(_killAndShutdown(killFuture));
+    return killFuture;
+  }
+
+  Future<void> _killAndShutdown(Future<int> killFuture) async {
+    unawaited(
+      killFuture.then((_) async {
+        await cancelAllStreamSubscriptions();
+      }),
+    );
+  }
+
+  Future<Map<String, dynamic>> waitFor({
+    String? event,
+    int? id,
+    Duration? timeout,
+    bool ignoreAppStopEvent = false,
+  }) async {
+    final Completer<Map<String, dynamic>> response =
+        Completer<Map<String, dynamic>>();
+    late StreamSubscription<String> sub;
+    sub = stdoutController.stream.listen((String line) async {
+      final dynamic json = _parseFlutterResponse(line);
```

Comment:
Is this type necessary? `_parseFlutterResponse` has a known, non-`dynamic` return type.

Reply:
probably not. All this code was written in pre-null safe Dart land and copied over in this PR. updated.